crates/shared/src/zeroex_api.rs

```diff
@@ -183,24 +369,66 @@ pub enum ZeroExResponseError {
 #[async_trait::async_trait]
 impl ZeroExApi for DefaultZeroExApi {
     async fn get_swap(&self, query: SwapQuery) -> Result<SwapResponse, ZeroExResponseError> {
-        self.swap_like_request_impl("quote", query).await
+        self.request(query.format_url(&self.base_url, "quote"))
+            .await
     }
 
     async fn get_price(&self, query: SwapQuery) -> Result<PriceResponse, ZeroExResponseError> {
-        self.swap_like_request_impl("price", query).await
+        self.request(query.format_url(&self.base_url, "price"))
+            .await
+    }
+
+    async fn get_orders_with_pagination(
+        &self,
+        query: &OrdersQuery,
+        results_per_page: usize,
+        page: usize,
+    ) -> Result<OrdersResponse, ZeroExResponseError> {
+        let mut url = query.format_url(&self.base_url);
+        url.query_pairs_mut()
+            .append_pair("page", &page.to_string())
+            .append_pair("perPage", &results_per_page.to_string());
+        self.request(url).await
+    }
+
+    async fn get_orders(
+        &self,
+        query: &OrdersQuery,
+    ) -> Result<Vec<OrderRecord>, ZeroExResponseError> {
+        let mut results = Vec::default();
+        let mut page = 1;
+        loop {
+            let mut response = self.get_orders_with_pagination(query, 100, page).await?;
+            let results_in_response = response.records.len();
+            results.append(&mut response.records);
+            if response.per_page != results_in_response as u64 {
```

Comment:
Wondering if it is possible (and meaningful) to add a UT to confirm this condition? (to make sure that if we send a request for which 0 records are returned that empty Vec is returned instead of some kind of error).

Reply:
I think testing `get_orders()` would require a mockable [`reqwest::Client`](https://docs.rs/reqwest_mock/latest/reqwest_mock/). This would require to replace `reqwest::Client` with `reqwest_mock::Client`. Either for this piece of code (but it would be weird to use 2 different `Client` types) or for the entire crate, which should then be done in a separate crate (if desired).

Without replacing the `Client` we could at least have a test to see how `get_orders_with_pagination()` behaves on empty results. This could be done easily by requesting something reasonable like `&page=10000000`.

Reply:
Yes, I was thinking about the latter one.

Reply:
Another way of doing it could be to factor out "handle_single_page" or something like that into a standalone function. As input it would the data the client downloaded and it would append to a mutable reference results vector and return whether there are more pages.
I often like this style of testing because there is less machinery to set up than with mocking.